File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current = head
        reverse = None;
        while current is not None:
            if reverse is None:
                reverse = ListNode(current.val, None)
            else:
                reverse = ListNode(current.val, reverse)
            current = current.next
            
        return reverse

File: Data Structures & Algorithms/reorder-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        ListNode newHead = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode prep = slow.next;
        // cut lose old connection
        slow.next = null;
        // start reverse swapNodes
        ListNode swapNode = null;
        while (prep != null) {
            ListNode next = prep.next;
            prep.next = swapNode;
            swapNode = prep;
            prep = next;
        }

        while (swapNode != null) {
            // store next value
            ListNode next = head.next;
            ListNode nextToSwap = swapNode.next;
            // operation
            head.next = swapNode;
            head = head.next;
            head.next = next;
            head = head.next;
            // cut connection
            swapNode = nextToSwap;
        }
    }
}
